pkg/cableengine/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,155 @@
+	Status  v1.ConnectionStatus
+	Latency *v1.LatencySpec
+}
+
+// NewEngine creates a new Engine for the local cluster
+func NewHealthChecker(localEndpointName string, client v1typed.GatewayInterface,
+	latencyMap *sync.Map) *HealthChecker {
+	return &HealthChecker{
+		localEndpointName: localEndpointName,
+		client:            client,
+		latencyMap:        latencyMap,
+	}
+}
+
+func (h *HealthChecker) Run(stopCh <-chan struct{}) {
+	go func() {
+		wait.Until(h.checkRemoteGatewayHealth, HealthCheckInterval, stopCh)
+	}()
+
+	klog.Infof("CableEngine healthchecker started")
+}
+func (h *HealthChecker) checkRemoteGatewayHealth() {
+	existingGw, err := h.client.Get(h.localEndpointName, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		klog.V(2).Infof("Gateway object not found: %v", err)
+		return
+	} else if err != nil {
+		klog.Errorf("Error retrieving the gateway object: %v", err)
+		return
+	}
+
+	if len(existingGw.Status.Connections) == 0 {
+		klog.V(2).Info("No endpoints present in the gateway")
+		return
+	}
+
+	deleteMap := copyLatencyMap(h.latencyMap)
+	var wg sync.WaitGroup
+
+	for _, endPointInfo := range existingGw.Status.Connections {
+		wg.Add(1)
+
+		hostName := endPointInfo.Endpoint.Hostname
+		healthCheckIp := endPointInfo.Endpoint.HealthCheckIP
+
+		// Remove the entry that is found in the gateway connection status
+		delete(deleteMap, hostName)
+
+		go func() {
+			latencyInfo := h.sendPing(&wg, hostName, healthCheckIp)
+			if latencyInfo != nil {
+				h.latencyMap.Store(hostName, latencyInfo)
+			}
+		}()
+	}
+
+	wg.Wait()
+
+	for k := range deleteMap {
+		// Remove the entry that are missing in the gateway connection status from the
+		// latencymap
+		h.latencyMap.Delete(k)
+	}
+}
+
+func copyLatencyMap(latencyMap *sync.Map) map[string]*LatencyInfo {
+	newMap := make(map[string]*LatencyInfo)
+
+	latencyMap.Range(func(key, val interface{}) bool {
+		if val != nil {
+			newMap[key.(string)] = val.(*LatencyInfo)
+		}
+		return true
+	})
+
+	return newMap
+}
+
+func (h *HealthChecker) sendPing(wg *sync.WaitGroup, remoteHostName, healthCheckIp string) *LatencyInfo {
+	defer wg.Done()
+	var lastRtt time.Duration
+
+	pinger, err := ping.NewPinger(healthCheckIp)
+	if err != nil {
+		klog.Errorf("Creating pinger for hostname %v to the ip %v failed due to %v", remoteHostName, healthCheckIp, err)
+		return nil
+	}
+
+	pinger.Count = 3
+	pinger.SetPrivileged(true)
+	pinger.Timeout = 3 * time.Second
+
+	pinger.OnRecv = func(packet *ping.Packet) {
+		lastRtt = packet.Rtt
+	}
+	var latencyinfo *LatencyInfo
+
+	pinger.OnFinish = func(stats *ping.Statistics) {
+		if stats.PacketLoss == 100 {
+			latencySpec := &v1.LatencySpec{
+				LastRTT:    0,
+				MinRTT:     0,
+				AverageRTT: 0,
+				MaxRTT:     0,
+				StdDevRTT:  0,
+			}
+			latencyinfo = &LatencyInfo{
+				Status:  v1.ConnectionError,
+				Latency: latencySpec,
+			}
+		} else {
+			latencySpec := &v1.LatencySpec{
+				LastRTT:    float64(lastRtt.Microseconds()) / 1000,
```

Comment:
```suggestion
				LastRTT:    float64(lastRtt.Microseconds()) / 1000,
```

I haven't looked but do they have .Milliseconds instead?, then you don't need to divide /1000

Reply:
The do have milli seconds , but to make the similar to the way the ping returns I just used floating point than an integer. And divided the microseconds.

Reply:
> > Some comments,
> > 
> > 1. I believe ping should be running in the background all the time to get meaningful statistics and reliable detection of the connection going down.
> > 2. Better integration to get the connection list from memory, or from the syncer that checks the connection status periodically instead of querying the k8s API for this.
> >    I will try if this is possible with the library and  refactor the code based on  the findings.

You sent no message :) I looked at the go-ping and probably will need some contribution to provide real-time statistics instead of just "at the end".

But I think it's easily doable. May be until then we can calculate the min/max/etc ... all by ourselves from the single packet reports.

Reply:
> The do have milli seconds , but to make the similar to the way the ping returns I just used floating point than an integer. And divided the microseconds.

Ack, ok, I expected them to expose float64 and not int. Makes sense then.

Reply:
now changed to unit64 as per the comments received for the API change.